staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -372,66 +398,164 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
-	slices = currentSlices
-
-	// Sort by name to ensure that keeping only the first slice is deterministic.
-	sort.Slice(slices, func(i, j int) bool {
-		return slices[i].Name < slices[j].Name
-	})
+	logger.V(5).Info("Existing slices", "obsolete", klog.KObjSlice(obsoleteSlices), "current", klog.KObjSlice(currentSlices))
 
 	if pool, ok := resources.Pools[poolName]; ok {
-		if pool.Generation > generation {
+		// Match each existing slice against the desired slices.
+		// Two slices match if they contain exactly the same
+		// device IDs, in an arbitrary order. Such a matched
+		// slice gets updated with the desired content if
+		// there is a difference.
+		//
+		// This supports updating the definition of devices
+		// in a slice. Adding or removing devices is done
+		// by deleting the old slice and creating a new one.
+		//
+		// This is primarily a simplification of the code:
+		// to support adding or removing devices from
+		// existing slices, we would have to identify "most
+		// similar" slices (= minimal editing distance).
+		//
+		// In currentSliceForDesiredSlice we keep track of
+		// which desired slice has a matched slice.
+		//
+		// At the end of the loop, each current slice is either
+		// a match or obsolete.
+		currentSliceForDesiredSlice := make(map[int]*resourceapi.ResourceSlice, len(pool.Slices))
+		for _, currentSlice := range currentSlices {
+			matched := false
+			for i := range pool.Slices {
+				desiredSlice := &pool.Slices[i]
+				if _, ok := currentSliceForDesiredSlice[i]; ok {
+					// Already has a match.
+					continue
+				}
+				if sameSlice(currentSlice, desiredSlice) {
+					currentSliceForDesiredSlice[i] = currentSlice
+					logger.V(5).Info("Matched existing slice", "slice", klog.KObj(currentSlice), "matchIndex", i)
+					matched = true
+					break
+				}
+			}
+			if !matched {
+				obsoleteSlices = append(obsoleteSlices, currentSlice)
+				logger.V(5).Info("Unmatched existing slice", "slice", klog.KObj(currentSlice))
+			}
+		}
+
+		// Desired metadata which must be set in each slice.
+		resourceSliceCount := len(pool.Slices)
+		desiredPool := resourceapi.ResourcePool{
+			Name:               poolName,
+			Generation:         generation, // May get updated later.
+			ResourceSliceCount: int64(resourceSliceCount),
+		}
+		desiredAllNodes := pool.NodeSelector == nil && nodeName == ""
+
+		// Now for each desired slice, figure out which of them are changed.
+		changedDesiredSlices := sets.New[int]()
+		for i, currentSlice := range currentSliceForDesiredSlice {
+			desiredSlice := &pool.Slices[i]
+
+			// Reordering entries is a difference and causes an update even if the
+			// entries are the same.
+			if !apiequality.Semantic.DeepEqual(&currentSlice.Spec.Pool, &desiredPool) ||
+				!apiequality.Semantic.DeepEqual(currentSlice.Spec.NodeSelector, pool.NodeSelector) ||
+				currentSlice.Spec.AllNodes != desiredAllNodes ||
+				!apiequality.Semantic.DeepEqual(currentSlice.Spec.Devices, desiredSlice.Devices) {
```

Comment:
`apiequality.Semantic.DeepEqual(currentSlice.Spec.Devices, desiredSlice.Devices)` Are there any order issues? Do we need to sort certain fields when comparing?

Reply:
If the order is different, the slices are not identical and need to be updated. From a semantic perspective that isn't relevant, but the controller is now specified as "creates slices exactly as intended by the driver", so order is relevant.

Reply:
got this.